### PR TITLE
Add logs to CPVM connection process

### PR DESCRIPTION
**Pull request by bernardodemarco**
### Description

The logs related to the CPVM connection are not enough to perform a great troubleshooting process.

This PR, therefore, adds logs and improves the existing ones in order to facilitate the CPVM connection troubleshooting. Furthermore, the legacy CPVM log proxy classes have been removed, in order to let the CPVM classes use log4j dependencies directly.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

### How Has This Been Tested?

- I created a VM and accessed it through a CPVM.
- Meanwhile, I verified that the messages were logged as expected.

### CODE REVIEW

File: services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVNCHandler.java [hunk 1]
```diff
@@ -155,31 +157,35 @@ public void onConnect(final Session session) throws IOException, InterruptedExce
...
     @OnWebSocketFrame
     public void onFrame(Frame f) throws IOException {
+        s_logger.trace("Sending client frame of {} bytes.", f.getPayloadLength());
         viewer.sendClientFrame(f);
     }
 
     @OnWebSocketError
     public void onError(Throwable cause) {
-        s_logger.error("Error on websocket", cause);
+        s_logger.error("Error on WebSocket.", cause);
```
Comment (DaanHoogland) [inline services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVNCHandler.java:189]:
```suggestion
        s_logger.error("Error on closing WebSocket.");
        s_logger.debug("Error on WebSocket.", cause);
```

File: services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVncClient.java [hunk 2]
```diff
@@ -72,9 +72,9 @@ public boolean isHostConnected() {
 
     @Override
     public boolean isFrontEndAlive() {
-        if (!connectionAlive || System.currentTimeMillis()
-                - getClientLastFrontEndActivityTime() > ConsoleProxy.VIEWER_LINGER_SECONDS * 1000) {
-            s_logger.info("Front end has been idle for too long");
+        long unusedTime = System.currentTimeMillis() - getClientLastFrontEndActivityTime();
+        if (!connectionAlive || unusedTime > ConsoleProxy.VIEWER_LINGER_SECONDS * 1000) {
+            s_logger.info("Front end has been idle for too long ({}ms).", unusedTime);
```
Comment (DaanHoogland) [inline services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVncClient.java:77]:
```suggestion
            s_logger.info("Front end has been idle for too long ({} ms).", unusedTime);
```

File: services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVncClient.java [hunk 3]
```diff
@@ -155,7 +156,7 @@ private boolean sendReadBytesToNoVNC(byte[] b, int readBytes) {
             session.getRemote().sendBytes(ByteBuffer.wrap(b, 0, readBytes));
             updateFrontEndActivityTime();
         } catch (WebSocketException | IOException e) {
-            s_logger.debug("Connection exception", e);
+            s_logger.error("VNC server connection exception.", e);
```
Comment (DaanHoogland) [inline services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVncClient.java:159]: stacktrace in `error`

File: services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVncClient.java [hunk 4]
```diff
@@ -300,8 +301,10 @@ private void connectClientToVNCServer(String tunnelUrl, String tunnelSession, St
                 ConsoleProxy.ensureRoute(getClientHostAddress());
                 client.connectTo(getClientHostAddress(), getClientHostPort());
             }
+
+            s_logger.info("Connection to VNC server has been established successfully.");
         } catch (Throwable e) {
-            s_logger.error("Unexpected exception", e);
+            s_logger.error("Unexpected exception while connecting to VNC server.", e);
```
Comment (DaanHoogland) [inline services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVncClient.java:307]:
```suggestion
            s_logger.error("Unexpected exception while connecting to VNC server, {}", e.getLocalizedMessage());
            s_logger.debug("Exception while connecting to VNC server.", e);
```

File: services/console-proxy/server/src/main/java/com/cloud/consoleproxy/vnc/NoVncClient.java [hunk 5]
```diff
@@ -175,8 +176,9 @@ public byte[] authenticateTunnel(String password)
                 is.readFully(buf);
                 String reason = new String(buf, RfbConstants.CHARSET);
 
-                s_logger.error("Authentication to VNC server is failed. Reason: " + reason);
-                throw new RuntimeException("Authentication to VNC server is failed. Reason: " + reason);
+                String msg = String.format("Authentication to VNC server has failed. Reason: %s", reason);
+                s_logger.error(msg);
```
Comment (DaanHoogland) [inline services/console-proxy/server/src/main/java/com/cloud/consoleproxy/vnc/NoVncClient.java:180]: :+1:

**Comment by bernardodemarco** [comment]
@DaanHoogland, thanks for your review!

As logging standards in CloudStack are still under discussion (#8746), I'd rather keep logging the exceptions stacktraces at the `error` level.

Honestly, I do not see the point in turning off the stacktraces at the `error` level. For operators that have a minimum software development background, the stacktraces can be a useful way to perform troubleshooting processes. 

On the other hand, if the operator has no knowledge of development, he might not understand the stacktraces, and that is fine, causing no harm.

Besides that, following along with what you've proposed in #8746, what happens when the operator chooses to turn the stacktraces down by setting the logs level to `error`? Are all exception logs be lost, since they are gonna be at `debug` level? If so, the difficulty of performing troubleshootings would increase significantly, as some errors/problems might be difficult to reproduce. 

Honestly, I'm looking forward to see the outcome of the #8746 discussion and I commit myself to participate on it. However, as I've mentioned before, I'd rather keep the proposed changes related to the stacktraces at the `error` level the way they are.

File: services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVNCHandler.java [hunk 6]
```diff
@@ -104,20 +105,20 @@ public void onConnect(final Session session) throws IOException, InterruptedExce
         try {
             port = Integer.parseInt(portStr);
         } catch (NumberFormatException e) {
-            logger.error("Invalid port value in query string: {}. Expected a number.", portStr, e);
+            logger.error(String.format("Invalid port value in query string: %s. Expected a number.", portStr), e);
```
Comment (DaanHoogland) [inline services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVNCHandler.java:108]: can you undo this change?

File: services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVNCHandler.java [hunk 7]
```diff
@@ -104,20 +105,20 @@ public void onConnect(final Session session) throws IOException, InterruptedExce
...
             throw new IllegalArgumentException(e);
         }
 
         if (ajaxSessionIdStr != null) {
             try {
                 ajaxSessionId = Long.parseLong(ajaxSessionIdStr);
             } catch (NumberFormatException e) {
-                logger.error("Invalid ajaxSessionId (sess) value in query string: {}. Expected a number.", ajaxSessionIdStr, e);
+                logger.error(String.format("Invalid ajaxSessionId (sess) value in query string: %s. Expected a number.", ajaxSessionIdStr), e);
```
Comment (DaanHoogland) [inline services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVNCHandler.java:116]: can you undo this change

File: services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVNCHandler.java [hunk 8]
```diff
@@ -147,7 +150,7 @@ public void onConnect(final Session session) throws IOException, InterruptedExce
             viewer = ConsoleProxy.getNoVncViewer(param, ajaxSessionIdStr, session);
             logger.debug("Viewer has been created successfully.");
         } catch (Exception e) {
-            logger.error("Failed to create viewer due to {}", e.getMessage(), e);
+            logger.error(String.format("Failed to create viewer due to %s", e.getMessage()), e);
```
Comment (DaanHoogland) [inline services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVNCHandler.java:153]: can you undo this change?

File: services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVNCHandler.java [hunk 9]
```diff
@@ -180,7 +182,7 @@ public void onClose(Session session, int statusCode, String reason) throws IOExc
 
     @OnWebSocketFrame
     public void onFrame(Frame f) throws IOException {
-        logger.trace("Sending client frame of {} bytes.", f.getPayloadLength());
+        logger.trace(String.format("Sending client frame of %s bytes.", f.getPayloadLength()));
```
Comment (DaanHoogland) [inline services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVNCHandler.java:185]: can you undo this change?

File: services/console-proxy/server/src/main/java/com/cloud/consoleproxy/vnc/NoVncClient.java [hunk 10]
```diff
@@ -79,15 +79,15 @@ public void connectTo(String host, int port, String path, String session, boolea
                 port = 80;
         }
 
-        logger.info("Connecting to VNC server {}:{} ...", host, port);
+        logger.info(String.format("Connecting to VNC server %s:%s ...", host, port));
```
Comment (DaanHoogland) [inline services/console-proxy/server/src/main/java/com/cloud/consoleproxy/vnc/NoVncClient.java:82]: can you undo this change?

File: services/console-proxy/server/src/main/java/com/cloud/consoleproxy/vnc/NoVncClient.java [hunk 11]
```diff
@@ -79,15 +79,15 @@ public void connectTo(String host, int port, String path, String session, boolea
...
         RawHTTP tunnel = new RawHTTP("CONNECT", host, port, path, session, useSSL);
         socket = tunnel.connect();
         setTunnelSocketStreams();
     }
 
     public void connectTo(String host, int port) {
         // Connect to server
-        logger.info("Connecting to VNC server {}:{} ...", host, port);
+        logger.info(String.format("Connecting to VNC server %s:%s ...", host, port));
```
Comment (DaanHoogland) [inline services/console-proxy/server/src/main/java/com/cloud/consoleproxy/vnc/NoVncClient.java:90]: can you undo this change?

File: services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVNCHandler.java (hunk 6 above)
Comment (bernardodemarco) [reply services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVNCHandler.java:108]: I'm not really sure why, but the `logger` fields of `com.cloud.consoleproxy.ConsoleProxyNoVNCHandler` and `com.cloud.consoleproxy.vnc.NoVncClient` are not instantiated directly from `log4j`. They're generated from a static method from the `com.cloud.consoleproxy.util.Logger` helper class.

The problem is that this helper class does not implement the new APIs provided by the new version of `log4j`, especially those that accept multiple parameters. As a consequence of that, `String.format()` must be used to log messages in those two classes.

IMO, it should be evaluated if the helper class is still needed. If so, we need to implement new methods to be able to use the new syntax provided by `log4j2`.

Comment (DaanHoogland) [reply services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVNCHandler.java:108]: I see, That was a good initiative back in 2013, to create a library facade, it had not been implemented fully in the project though;
```
// logger facility for dynamic switch between console logger used in Applet and log4j based logger
```
This might be an oversight in the log4j migration. `com.cloud.consoleproxy.ConsoleProxyLoggerFactory` is still using log4j v1 classes under the hood.

I personally think we should have implemented such a proxy library, for the whole project and implemented log4j under the hood. I think for now we will have to do with 'fixing' the console proxy classes.

**Review by DaanHoogland — APPROVED** [review]
clgtm, can we now remove `com.cloud.consoleproxy.util.Logger` from the code base @bernardodemarco ? Just to deal with the technical debt. (we can always add a new proxy once we get wise ;)

**Comment by bernardodemarco** [comment]
> clgtm, can we now remove `com.cloud.consoleproxy.util.Logger` from the code base @bernardodemarco ? Just to deal with the technical debt. (we can always add a new proxy once we get wise ;)

@DaanHoogland, I have just removed the CPVM log proxy classes and updated the PR's description.

**Comment by DaanHoogland** [comment]
> > clgtm, can we now remove `com.cloud.consoleproxy.util.Logger` from the code base @bernardodemarco ? Just to deal with the technical debt. (we can always add a new proxy once we get wise ;)
> 
> @DaanHoogland, I have just removed the CPVM log proxy classes and updated the PR's description.

thanks for honouring my obnoxious request @bernardodemarco .

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test alma9 kvm-alma9

**Review by hsato03 — APPROVED** [review]
LGTM. 

Tested using the web consoles and monitoring the CPVM logs at the same time. The logs are being created correctly.

**Comment by JoaoJandre** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test
